File: code/cpp/shortestPathDAG.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <limits>

#include "helpers.h"


std::vector<int> topsort(networks::Graph &graph);
std::pair<std::vector<int>, int> shortestPathDAG(networks::Graph &graph, int start, int target);
int dfs(int i, int node, std::vector<bool> &visited, std::vector<int> &visitedNodes, networks::Graph &graph);


int main()
{   
#ifndef FILE_INPUT
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif

    int numNodes, numEdges;
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    std::cin >> weighted >> directed;

    int u, v, w;
    std::vector<networks::Edge> edges;
    if (weighted)
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v >> w;
            edges.push_back(networks::Edge(u, v, w));
        }
    }
    else
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v;
            edges.push_back(networks::Edge(u, v));
        }
    }

    networks::Graph graph = networks::Graph(numNodes, edges, directed);

    int target = graph.numNodes-1;
    std::pair<std::vector<int>, int> ans = shortestPathDAG(graph, 0, target); 
    helpers::printVector(ans.first, "The Shortest path Array in the Given DAG is: ");
    std::cout << "The shortest path to " << target << " is " << ans.second;

    return 0;
}


std::vector<int> topsort(networks::Graph &graph)
{
    int n = graph.numNodes;
    std::vector<bool> visited(n, false);
    std::vector<int> ordering(n);
    int i= n-1;

    for (int node=0; node<n; node++)
    {
        if (!visited[node])
            i = dfs(i, node, visited, ordering, graph);
        
    }

    return ordering;    
}

int dfs(int i, int node, std::vector<bool> &visited, std::vector<int> &ordering, networks::Graph &graph)
{
    visited[node] = true;

    std::vector<std::pair<int, int>> edges = graph.adjList[node];
    for (auto edge : edges)
    {
        if (!visited[edge.first])
            i = dfs(i, edge.first, visited, ordering, graph);
    }

    ordering[i] = node;
    return i-1;
}

std::pair<std::vector<int>, int> shortestPathDAG(networks::Graph &graph, int start, int target)
{
    if (target == -1) target = graph.numNodes-1;

    std::vector<int> ordering = topsort(graph);
    std::vector<int> distance(graph.numNodes, INT32_MAX);
    distance[start] = 0;

    for (int i=0; i < graph.numNodes; i++)
    {
        int nodeIndex = ordering[i];
        std::vector<std::pair<int, int>> adjEdges = graph.adjList[nodeIndex];
        for (auto edge : adjEdges)
        {
            int newDist = distance[nodeIndex] + edge.second;
            if (distance[edge.first] == INT32_MAX) distance[edge.first] = newDist;
            else distance[edge.first] = std::min(distance[edge.first], newDist);
        }
    }

    std::pair<std::vector<int>, int> ans = {distance, distance[target]};
    return ans;
}

File: code/python/shortest_path_dag.py
import helpers as H

# global variables
INPUT_PATH = 'input.txt'
OUTPUT_PATH = 'output.txt'

# clearing the contents of the file if any
open(OUTPUT_PATH, 'w').close()

def topsort(graph: H.Graph):
    n = graph.num_nodes
    visited = [False for _ in range(n)]
    ordering = [0 for _ in range(n)]
    i = n-1
    
    for node in range(n):
        if not visited[node]:
            wrapper = [visited]  # making a wrapper to pass by reference
            i = dfs(i, node, wrapper, ordering, graph)
            
    return ordering
    
            
def dfs(i, node, v_wrapper, ordering, graph: H.Graph):
    v_wrapper[0][node] = True
    edges = graph.adj_list[node]
    for edge in edges:
        if not v_wrapper[0][edge[0]]:
             i = dfs(i, edge[0], v_wrapper, ordering, graph)
             
    ordering[i] = node
    return i-1


def shortest_path_dag(graph: H.Graph, start, target=-1):
    if target == -1: target = graph.num_nodes - 1
    
    ordering = topsort(graph)
    distance = [float('inf') for _ in range(graph.num_nodes)]
    distance[start] = 0
    
    for i in range(graph.num_nodes):
        node_index = ordering[i]
        adj_edges = graph.adj_list[node_index]
        for edge in adj_edges:
            new_dist = distance[node_index] + edge[1]
            if distance[edge[0]] == float('inf'): distance[edge[0]] = new_dist 
            else: distance[edge[0]] = min(distance[edge[0]], new_dist)
            
    return distance, distance[target]
    
    
if __name__ == '__main__':
    with open(INPUT_PATH, 'r') as f:
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        edges = []
        if weighted:
            for i in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v, w))
        else:
            for i in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v))
                
    graph = H.Graph(num_nodes, edges, directed)
    
    ans = shortest_path_dag(graph, 0)
    with open(OUTPUT_PATH, 'a') as f:
        f.write(f"{ans}")